src/OpenTelemetry.Extensions/Trace/AutoFlushActivityProcessor.cs

```diff
@@ -0,0 +1,60 @@
+﻿// <copyright file="AutoFlushActivityProcessor.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// Activity processor that flushes its containing <see cref="TracerProvider"/> if an ended
+    /// activity matches a predicate.
+    /// </summary>
+    public sealed class AutoFlushActivityProcessor : BaseProcessor<Activity>
+    {
+        private readonly Predicate<Activity> predicate;
+        private readonly int timeoutMilliseconds;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoFlushActivityProcessor"/> class.
+        /// </summary>
+        /// <param name="predicate">Predicate that should return <c>true</c> to initiate a flush.</param>
+        /// <param name="timeoutMilliseconds">Timeout (in milliseconds) to use for flushing.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the <c>timeoutMilliseconds</c> is smaller than -1.
+        /// </exception>
+        public AutoFlushActivityProcessor(Predicate<Activity> predicate, int timeoutMilliseconds)
+        {
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            if (timeoutMilliseconds < Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds));
+            }
+
+            this.timeoutMilliseconds = timeoutMilliseconds;
+        }
+
+        /// <inheritdoc/>
+        public override void OnEnd(Activity data)
+        {
+            if (this.predicate(data))
```

Comment:
What if the predicate throws?

Reply:
IMHO predicate's exception should be bubbled to the OnEnd caller as this class is not defining the predicate code. Otherwise it can be swallowed and logged with debug level.

Reply:
What should the user expect (e.g. if a user provides a buggy predicate that throws, should he expect the processor to fail and the data to be dropped, or the predicate itself to be ignore - so no Flush is called but the data still flows, or some other behavior?)

Reply:
I agree. Data should still flow. Wrapped the predicate with `try...catch` block and logged exception.

Reply:
What does it mean for the data (not) to "flow"? Are expectations if a Span Processor's OnEnd throws an exception documented? I think we should not make any additional guarantees / set any additional expectations here.

Reply:
The way I imagine: "if the predicate throws exception, the `AutoFlushActivityProcessor` will swallow the exception and treat the predicate as False (which means `ForceFlush` will not be applied)".

Reply:
I stick with "swallow and log" approach and mentioned this behaviour in the doc.